compiler/rustc_type_ir/src/elaborate.rs

```diff
@@ -179,16 +179,39 @@ impl<I: Interner, O: Elaboratable<I>> Elaborator<I, O> {
                     ),
                 };
             }
-            // `T: [const] Trait` implies `T: [const] Supertrait`.
-            ty::ClauseKind::HostEffect(data) => self.extend_deduped(
-                cx.explicit_implied_const_bounds(data.def_id()).iter_identity().map(|trait_ref| {
-                    elaboratable.child(
-                        trait_ref
-                            .to_host_effect_clause(cx, data.constness)
-                            .instantiate_supertrait(cx, bound_clause.rebind(data.trait_ref)),
-                    )
-                }),
-            ),
+            ty::ClauseKind::HostEffect(data) => {
+                // `T: [const] Trait` implies `T: [const] Supertrait`.
+                self.extend_deduped(
+                    cx.explicit_implied_const_bounds(data.def_id()).iter_identity().map(
+                        |trait_ref| {
+                            elaboratable.child(
+                                trait_ref
+                                    .to_host_effect_clause(cx, data.constness)
+                                    .instantiate_supertrait(
+                                        cx,
+                                        bound_clause.rebind(data.trait_ref),
+                                    ),
+                            )
+                        },
+                    ),
+                );
+
+                // `Adt: [const] Trait` implies each field also implements `[const] Trait`
```

Comment:
If we did apply this, I'd probably want to abstract out the function that maps from a ty to its constituent types and share it between the `Destruct` trait logic and here.

Logic would probably need to be pulled out of https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_next_trait_solver/solve/assembly/structural_traits.rs.html#743-820